index.html

```diff
@@ -1603,6 +1622,348 @@ <h2>
+    </caption>
+    <tr>
+      <th>
+        Parameter name
+      </th>
+      <th>
+        Value type
+      </th>
+      <th>
+        Required
+      </th>
+    </tr>
+    <tr>
+      <td>
+        type
+      </td>
+      <td>
+        String
+      </td>
+      <td>
+        yes
+      </td>
+    </tr>
+    <tr>
+      <td>
+        sampleInterval
+      </td>
+      <td>
+        Number
+      </td>
+      <td>
+        yes
+      </td>
+    </tr>
+    <tr>
+      <td>
+        samples
+      </td>
+      <td>
+        Array of strings
+      </td>
+      <td>
+        yes
+      </td>
+    </tr>
+  </table>
+  <p>
+    The [=remote end steps=]  given |session|, |URL variables| and |parameters| are:
+  </p>
+  <ol class="algorithm" data-cite="WEBDRIVER2">
+    <li>
+      Let |virtualPressureSourceType| be the result of invoking
+      <a data-cite="!WEBDRIVER2#dfn-getting-properties">get a property</a> "type" from |parameters|.
+    </li>
+    <li>
+      If |virtualPressureSourceType| is not a [=string=], return [=error=] with [=error code|WebDriver error code=] [=invalid argument=].
+    </li>
+    <li>
+      Let |sampleInterval| be the result of invoking
+      <a data-cite="!WEBDRIVER2#dfn-getting-properties">get a property</a> "sampleInterval" from |parameters|.
+    </li>
+    <li>
+      If |sampleInterval| is not an Number, return [=error=] with [=error code|WebDriver error code=] [=invalid argument=].
+    </li>
+    <li>
+      Let |samples| be the result of invoking
+      <a data-cite="!WEBDRIVER2#dfn-getting-properties">get a property</a> "samples" from |parameters|.
+    </li>
+    <li>
+      If |samples| is not an Array, return [=error=] with [=error code|WebDriver error code=] [=invalid argument=].
+    </li>
+    <li>
+      For each |sample| in |samples|:
+      <ol>
+        <li>
+          If |sample| is not a [=string=] and a valid {{PressureState}}, return [=error=] with [=error code|WebDriver error code=] [=invalid argument=].
+        </li>
+        <li>
+          [=queue/Enqueue=] |sample| to |virtualPressureSource|'s [=pending samples=].
+        </li>
+      </ol>
+    </li>
+    <li>
+      Let |topLevelTraversable| be the <a data-cite="webdriver2/#dfn-current-browsing-context">current browsing
+      context</a>'s [=browsing context/top-level traversable=].
+    </li>
+    <li>
+      Let |topLevelVirtualPressureSourceMapping| be the |topLevelTraversable|'s [=virtual pressure source mapping=].
+    </li>
+    <li>
+      If |topLevelVirtualPressureSourceMapping| contains |virtualPressureSourceType|, return [=error=] with
+      [=error code|WebDriver error code=] [=invalid argument=].
+    </li>
+    <li>
+      Let |virtualPressureSource| be a new [=virtual pressure source=].
+    </li>
+    <li>
+      Let |virtualPressureSource|'s [=sample update interval=] be |sampleInterval|.
+    </li>
+    <li>
+      Run the following steps [=in parallel=] at the |sampleInterval| time interval:
```

Comment:
I can rewrite it to use that, yes, but I am wondering what makes the most sense.

It is kind of nice if the constuctor function (this endpoint) sets that up, but then when there is no new data it will still run and just do nothing. I don't think that is a problem, but some might. If we start creating timers or recreating when we update with new data, then there is extra case when pushing new data before it is done with the existing ones. 

Are you OK with this former approach?

Reply:
> One option to tackle both is to leave more of it up to users. For example:
> 
> * Take one update value in "update virtual pressure source" and invoke the data collection steps for all globals belonging to a given top-level traversable.
> * Users (including devtools-frontend in Chromium's case) can then implement whatever logic they want to call "update virtual pressure source" with different values at a rate they want.
> 
> This effectively makes the sampling rate up to the users in the WebDriver case, which I'm not sure is something desirable or not. It'd make the spec simpler, though.

We ended up adopting this approach.